daemon/graphdriver/projectquota.go

```diff
@@ -0,0 +1,292 @@
+		uintptr(unsafe.Pointer(&d)), 0, 0)
+	if errno != 0 {
+		return fmt.Errorf("Failed to get quota limit for projid %d on %s: %v",
+			projectID, q.backingFsDev, errno.Error())
+	}
+	quota.Size = uint64(d.d_blk_hardlimit) * 512
+
+	return nil
+}
+
+func getProjectID(targetPath string) (uint32, error) {
+	dir, err := openDir(targetPath)
+	if err != nil {
+		return 0, err
+	}
+	defer closeDir(dir)
+
+	var fsx C.struct_fsxattr
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSGETXATTR,
+		uintptr(unsafe.Pointer(&fsx)))
+	if errno != 0 {
+		return 0, fmt.Errorf("Failed to get projid for %s: %v", targetPath, errno.Error())
+	}
+
+	return uint32(fsx.fsx_projid), nil
+}
+
+func setProjectID(targetPath string, projectID uint32) error {
+	dir, err := openDir(targetPath)
+	if err != nil {
+		return err
+	}
+	defer closeDir(dir)
+
+	var fsx C.struct_fsxattr
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSGETXATTR,
+		uintptr(unsafe.Pointer(&fsx)))
+	if errno != 0 {
+		return fmt.Errorf("Failed to get projid for %s: %v", targetPath, errno.Error())
+	}
+	fsx.fsx_projid = C.__u32(projectID)
+	fsx.fsx_xflags |= C.FS_XFLAG_PROJINHERIT
+	_, _, errno = syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSSETXATTR,
+		uintptr(unsafe.Pointer(&fsx)))
+	if errno != 0 {
+		return fmt.Errorf("Failed to set projid for %s: %v", targetPath, errno.Error())
+	}
+
+	return nil
+}
+
+func getMaxProjectID(path string) (uint32, error) {
+	//
+	// Get project id of parent dir as test for project quota support
+	//
+	maxProjectID, err := getProjectID(path)
+	if err != nil {
+		return 0, err
+	}
+	if maxProjectID < 1 {
+		maxProjectID = 1
+	}
+
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return 0, fmt.Errorf("read directory failed : %s", path)
+	}
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+		projid, err := getProjectID(filepath.Join(path, file.Name()))
+		if err != nil {
+			return 0, err
+		}
+		if maxProjectID <= projid {
+			maxProjectID = projid + 1
+		}
+	}
+
+	logrus.Debugf("getMaxProjectID(%s) = %d", path, maxProjectID)
+	return maxProjectID, nil
+}
+
+func free(p *C.char) {
+	C.free(unsafe.Pointer(p))
+}
+
+func openDir(path string) (*C.DIR, error) {
+	Cpath := C.CString(path)
+	defer free(Cpath)
+
+	dir := C.opendir(Cpath)
+	if dir == nil {
+		return nil, fmt.Errorf("Can't open dir")
+	}
+	return dir, nil
+}
+
+func closeDir(dir *C.DIR) {
+	if dir != nil {
+		C.closedir(dir)
+	}
+}
+
+func getDirFd(dir *C.DIR) uintptr {
+	return uintptr(C.dirfd(dir))
+}
+
+// get the backing block device of the driver home directory
+// and create a block device node under the home directory
+// to be used by quotactl commands
+func makeBackingFsDev(home string) (string, error) {
+	fileinfo, err := os.Stat(home)
+	if err != nil {
+		return "", err
+	}
+
+	backingFsDev := path.Join(home, "backingFsDev")
+	syscall.Unlink(backingFsDev)
```

Comment:
It is just a guarantee. The way xfs_quota tool does it to get the xfs mount point as argument and deduce the the block device from /proc/mounts. So having the backingFsDev actual node (as a duplicate of the original node (e.g. /dev/dm-1) is just a matter of convenience and less complex code.